pkg/utils/bpf-iter-ns/bpf-iter-ns.go

```diff
@@ -0,0 +1,197 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package bpfiterns reads a ebpf iterator in a different namespace.
+package bpfiterns
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/cilium/ebpf/link"
+	systemdDbus "github.com/coreos/go-systemd/v22/dbus"
+	_ "github.com/godbus/dbus/v5"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/utils/host"
+)
+
+func init() {
+	rand.Seed(time.Now().Unix())
+}
+
+func IsHostPidNs() (bool, error) {
+	selfFileInfo, err := os.Stat("/proc/self/ns/pid")
+	if err != nil {
+		return false, err
+	}
+	selfStat, ok := selfFileInfo.Sys().(*syscall.Stat_t)
+	if !ok {
+		return false, errors.New("not a syscall.Stat_t")
+	}
+
+	systemdFileInfo, err := os.Stat(fmt.Sprintf("%s/1/ns/pid", host.HostProcFs))
+	if err != nil {
+		return false, err
+	}
+	systemdStat, ok := systemdFileInfo.Sys().(*syscall.Stat_t)
+	if !ok {
+		return false, errors.New("not a syscall.Stat_t")
+	}
+
+	return selfStat.Ino == systemdStat.Ino, nil
+}
+
+// Read reads the iterator in the host pid namespace.
+// It will test if the current pid namespace is the host pid namespace.
+func Read(iter *link.Iter) ([]byte, error) {
+	isHostPidNs, err := IsHostPidNs()
+	if err != nil {
+		return nil, fmt.Errorf("check if host pid namespace: %w", err)
+	}
+
+	if isHostPidNs {
+		return ReadOnCurrentPidNs(iter)
+	} else {
+		return ReadOnHostPidNs(iter)
+	}
+}
+
+// ReadOnCurrentPidNs reads the iterator in the current pid namespace.
+func ReadOnCurrentPidNs(iter *link.Iter) ([]byte, error) {
+	file, err := iter.Open()
+	if err != nil {
+		return nil, fmt.Errorf("open BPF iterator: %w", err)
+	}
+	defer file.Close()
+	buf, err := io.ReadAll(file)
+	if err != nil {
+		return nil, fmt.Errorf("read BPF iterator: %w", err)
+	}
+	return buf, err
+}
+
+// ReadOnHostPidNs reads the iterator in the host pid namespace.
+// It does so by pinning the iterator in a temporary directory in the host bpffs,
+// and then creating a systemd service that will read the iterator and write it
+// to a temporary pipe. The pipe is then read and returned.
+func ReadOnHostPidNs(iter *link.Iter) ([]byte, error) {
+	selfPidOnHost, err := os.Readlink(host.HostProcFs + "/self")
+	if err != nil {
+		return nil, fmt.Errorf("readlink /proc/self: %w", err)
+	}
+	if selfPidOnHost == "" {
+		return nil, fmt.Errorf("empty /proc/self symlink")
+	}
+
+	tmpPinDir, err := os.MkdirTemp("/sys/fs/bpf", "ig-iter-")
```

Comment:
Shouldn't it prepend `host.HostProcFs` un case /sys/fs/bpf is not mounted?

Reply:
Yes (but host.HostRoot, not host.HostProcFs)